### PR TITLE
feat: similarity_search() for LangChain + upsert_relations() idempotency test

### LINKED ISSUE #20: fix: vector_search() always returns empty — E2E tests + LangChain similarity_search()
## Problem

`CoordinodeClient.vector_search()` always returns an empty list, even when nodes with `embedding` properties exist in the graph. This silently breaks:

- `CoordinodePropertyGraphStore.vector_query()` in llama-index adapter
- Any LangChain workflow relying on vector similarity

```python
client = CoordinodeClient("localhost:7080")
results = client.vector_search(label="Chunk", property="embedding", vector=[...], top_k=5)
# → [] (always empty, even with data)
```

## Root cause

This is a **DB-side gap**: `VectorService` gRPC endpoint exists and accepts requests, but the wiring from VectorService to the HNSW index executor is incomplete in CoordiNode v0.3. The RPC returns without querying the index.

Tracked as **G073** in CoordiNode DB repository.

## SDK work (after G073 is fixed in DB)

Once G073 is resolved and `vector_search` returns actual results:

1. **Add E2E integration test** — `test_vector_search_returns_results` with a real embedding vector
2. **Add `similarity_search()` to LangChain adapter** (`CoordinodeGraph`) — currently missing; should wrap `client.vector_search()` with label/property defaults
3. **Verify `vector_query()` in LlamaIndex adapter** — test against live DB, confirm scores are correct (currently returns `r.distance` — verify this matches expected similarity semantics)

## Acceptance criteria

- [ ] `client.vector_search(label="Chunk", property="embedding", vector=[...], top_k=5)` returns results when data exists
- [ ] Integration test passes against live CoordiNode instance
- [ ] `CoordinodeGraph.similarity_search(query_vector, k=5)` implemented in LangChain adapter
- [ ] LlamaIndex `vector_query()` tested E2E

## Gate

**Blocked by G073** — requires CoordiNode DB fix: VectorService → HNSW executor wiring.
Do not start SDK implementation until a CoordiNode release with G073 fixed is available.

### LINKED ISSUE #21: fix: duplicate edges on re-ingestion — replace CREATE with MERGE in upsert_relations() and add_graph_documents()
## Problem

Re-ingesting the same data creates duplicate edges. Every call to `upsert_relations()` (LlamaIndex) or `add_graph_documents()` (LangChain) creates new edges unconditionally:

```python
# First ingest
store.upsert_relations([Relation(label="KNOWS", source_id="alice", target_id="bob")])
# Second ingest — creates a DUPLICATE edge
store.upsert_relations([Relation(label="KNOWS", source_id="alice", target_id="bob")])
```

This makes repeated indexing non-idempotent and bloats the graph.

## Root cause

CoordiNode Cypher does not support `MERGE` for relationship patterns — only for node patterns (NodeScan). The adapters work around this with `CREATE`, accepting duplicate edges as a known limitation.

Tracked as **G072** in CoordiNode DB repository: `MERGE (src)-[r:TYPE]->(dst)` fails with "MERGE create from non-NodeScan pattern".

## SDK changes (after G072 is fixed in DB)

Once G072 is resolved:

**`llama-index-coordinode/llama_index/graph_stores/coordinode/base.py`**
- `upsert_relations()`: replace `CREATE (src)-[r:{label}]->(dst)` → `MERGE (src)-[r:{label}]->(dst)`
- Remove the comment block explaining the CREATE fallback
- Remove the `SET r += $props` workaround comment

**`langchain-coordinode/langchain_coordinode/graph.py`**
- `_create_edge()`: replace `CREATE (src)-[r:{rel_type}]->(dst)` → `MERGE (src)-[r:{rel_type}]->(dst)`
- Rename `_create_edge()` → `_upsert_edge()` (or keep name, update docstring)
- `_link_document_to_entities()`: replace `CREATE (d)-[:MENTIONS]->(n)` → `MERGE (d)-[:MENTIONS]->(n)`
- `add_graph_documents()` docstring: remove duplicate-edge warning

**Tests**
- `test_add_graph_documents_idempotent`: change assertion from `cnt >= 1` to `cnt == 1`
- Add similar idempotency test for LlamaIndex `upsert_relations()`

## Acceptance criteria

- [ ] Re-ingesting the same document creates exactly 1 edge (not N)
- [ ] `upsert_relations()` is idempotent
- [ ] `add_graph_documents()` is idempotent for both nodes and edges
- [ ] All existing tests pass

## Gate

**Blocked by G072** — requires CoordiNode DB fix: MERGE clause for relationship patterns in Cypher executor.
Do not start SDK implementation until a CoordiNode release with G072 fixed is available.

### PR DESCRIPTION
**Pull request by polaz**
## Summary

- Add `similarity_search(query_vector, k, label, property)` to `CoordinodeGraph` (LangChain adapter), wrapping `CoordinodeClient.vector_search()` with label/property defaults
- Add `test_upsert_relations_idempotent` integration test for LlamaIndex adapter, verifying exactly 1 edge after double `upsert_relations()` call (MERGE semantics)

## Technical Details

**`similarity_search()`**:
- Returns `[{"id": ..., "node": ..., "distance": ...}, ...]` sorted by ascending cosine distance
- Guards against empty `query_vector` (server returns `INVALID_ARGUMENT` for empty vectors)
- Two integration tests: seeded `:LCSim` node found in top-k results, empty vector returns `[]`

**Idempotency test**:
- Seeds src/dst nodes, calls `upsert_relations()` twice with the same `Relation`
- Verifies `count(r) == 1` via Cypher — MERGE semantics confirmed

## Test Plan

- 55/55 integration tests pass against live CoordiNode instance

Closes #20
Closes #21

### CODE REVIEW

File: langchain-coordinode/langchain_coordinode/graph.py [hunk 1]
```diff
@@ -194,6 +194,39 @@ def query(
...
+        """
+        if not query_vector:
+            return []
+        results = self._client.vector_search(
+            label=label,
+            property=property,
+            vector=query_vector,
+            top_k=k,
+        )
+        return [{"id": r.node.id, "node": r.node.properties, "distance": r.distance} for r in results]
```
[the remark this answers is not in the document]
Comment (polaz) [reply langchain-coordinode/langchain_coordinode/graph.py]: @coderabbitai Fixed — results are now sorted client-side by `r.distance` ascending before the list comprehension.

File: tests/integration/adapters/test_langchain.py [hunk 2]
```diff
@@ -133,6 +133,44 @@ def test_add_graph_documents_idempotent(graph, unique_tag):
...
+        assert results[0]["distance"] >= 0.0
+    finally:
+        graph.query("MATCH (n:LCSim {id: $id}) DELETE n", params={"id": f"lcsim-{unique_tag}"})
+
+
+def test_similarity_search_empty_vector_returns_empty(graph):
+    """similarity_search() with an empty vector list returns an empty list without error."""
+    results = graph.similarity_search([], k=5)
+    assert isinstance(results, list)
+
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/adapters/test_langchain.py:176]: @coderabbitai Fixed — assertion changed to `results == []` to explicitly enforce the empty-vector contract.

File: tests/integration/adapters/test_langchain.py [hunk 3]
```diff
@@ -133,6 +133,44 @@ def test_add_graph_documents_idempotent(graph, unique_tag):
...
+        )
+        seeded_internal_id = seed_rows[0]["nid"]
+
+        results = graph.similarity_search(vec, k=5, label="LCSim", property="embedding")
+
+        assert isinstance(results, list)
+        assert len(results) >= 1
+        assert all("id" in r and "node" in r and "distance" in r for r in results)
+        assert any(r["id"] == seeded_internal_id for r in results)
+        assert results[0]["distance"] >= 0.0
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/adapters/test_langchain.py]: @coderabbitai Won't fix — `graph.query()` wraps `CoordinodeClient.cypher()` which returns raw dict values. CoordiNode's `CREATE (n:...) RETURN n` yields the internal integer node ID directly (not a node object with `.id_`). Calling `.id_` on it would raise `AttributeError`. `similarity_search()` returns `{"id": r.node.id, ...}` (also integer), so direct equality comparison is correct and the test passes. Added an inline comment to clarify this.

File: langchain-coordinode/langchain_coordinode/graph.py [hunk 4]
```diff
@@ -194,6 +194,45 @@ def query(
         # cypher() returns List[Dict[str, Any]] directly — column name → value.
         return self._client.cypher(query, params=params or {})
 
+    def similarity_search(
+        self,
+        query_vector: list[float],
+        k: int = 10,
```
[the remark this answers is not in the document]
Comment (polaz) [reply langchain-coordinode/langchain_coordinode/graph.py]: @coderabbitai Fixed — changed to `k: int = 5` to align with issue #20 acceptance criteria.

File: tests/integration/adapters/test_langchain.py [hunk 5]
```diff
@@ -133,6 +133,48 @@ def test_add_graph_documents_idempotent(graph, unique_tag):
...
+        # where r.node.id is the same integer. Direct equality comparison is correct.
+        seeded_internal_id = seed_rows[0]["nid"]
+
+        results = graph.similarity_search(vec, k=5, label="LCSim", property="embedding")
+
+        assert isinstance(results, list)
+        assert len(results) >= 1
+        assert all("id" in r and "node" in r and "distance" in r for r in results)
+        assert any(r["id"] == seeded_internal_id for r in results)
+        assert results[0]["distance"] >= 0.0
```
[the remark this answers is not in the document]
Comment (polaz) [reply tests/integration/adapters/test_langchain.py]: @coderabbitai Fixed — replaced non-negativity check with `assert all(results[i]["distance"] <= results[i+1]["distance"] for i in range(len(results) - 1))` which validates the ascending-order contract.
